### PR TITLE
[Dynamic Spec]Add ObjectSpec

### PR DESCRIPTION
**Pull request by fxdawnn**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #184129
* #184299
* #184271
* #184440
* __->__ #182764

## Summary

Adds `ObjectSpec` and extends `lookup_spec_from_dynamo_source` so the spec system can address tensors reached via Python attribute access (`obj.weight`, `nn.Module` params/buffers, `self.x` inside an instance method).

```python
import torch
from torch.fx.experimental.dynamic_spec import (
    ObjectSpec, ParamsSpec, ShapesSpec, ShapeVar, TensorSpec,
)

class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.weight = torch.nn.Parameter(torch.randn(4, 3))
    def forward(self):
        return self.weight + 1

torch.compile(
    Model(),
    shapes_spec=ShapesSpec(
        params=ParamsSpec(
            {"self": ObjectSpec({"weight": TensorSpec([ShapeVar("h"), None])})}
        )
    ),
)
```

### `torch/fx/experimental/dynamic_spec.py`

- **`ObjectSpec`** — attribute-keyed container.
  - Constructor: `ObjectSpec({name: IntermediateSpec, ...})`. Values may be leaves (`TensorSpec` / `IntVar` / `int` / `None`) or another `ObjectSpec` for recursion.
  - Implements `__getitem__` / `__contains__` / `__iter__` / `__len__` / `items()` for dict-like access.
  - `__repr__` matches the module's existing style (`object_spec:\n  .name: ...`); nested children indent recursively.
  - `to_jsonable()` mirrors the pattern used by other spec types — recurses into spec children, passes raw leaves through.
- **`IntermediateSpec`** type alias updated to `LeafSpec | ObjectSpec` so containers can hold mixed leaf and nested-spec values.

### `torch/_dynamo/variables/builder.py`

- **`lookup_spec_from_dynamo_source`** — extended to walk `Source` chains:
  - `LocalSource(name, is_input=True)` — the root of any walk.
  - `AttrSource(base, member)` — descends `ObjectSpec._fields[member]`.
  - `NNModuleSource` (and subclasses) — transparently unwrapped (guard-semantics marker, not an access step).
  - `DictGetItemSource(UnspecializedParamBufferSource(_, '_parameters' | '_buffers'), key)` — dynamo internally rewrites `self.weight` as `self._parameters["weight"]`; the walk collapses that pair into a single `("attr", key)` step so the user-facing attribute name matches the spec.
  - Other source kinds return `None` — later container PRs (DictSpec / ListSpec) extend this dispatch.

- **`wrap_module`** — the pre-marking loop that calls `mark_static_input` on each `named_parameter` / `named_buffer` now skips entries with a spec. Without this, the static stamp would be applied before `wrap_tensor`'s spec-aware bypasses get a chance to fire, and the Parameter would still be lifted as a graph attribute.

Combined effect: `nn.Parameter` attributes participate in spec-driven dynamism end-to-end.

## Test plan — `python test/dynamo/test_dynamic_spec.py`

- **`TestObjectSpec`** (data class) — empty / dict-construction / iter+items / recursive nesting / repr (none-leaf / nested-tensor / nested-objectspec) / `to_jsonable`.
- **`TestObjectSpecLookup`** (lookup walk, ordered simplest → most complex):
  - `test_local_source_root_returns_top_level_spec` — bare `LocalSource` returns the `ObjectSpec` itself.
  - `test_attr_descends_into_objectspec` — `AttrSource(LocalSource(...), "weight")`.
  - `test_nested_objectspec_walk` — three-level `model.inner.weight`.
  - `test_missing_attr_returns_none` — attr not present in `ObjectSpec`.
  - `test_attr_against_non_objectspec_returns_none` — type mismatch (top-level is `TensorSpec`, source asks for an attr).
  - `test_nn_module_source_is_unwrapped` — `AttrSource(NNModuleSource(LocalSource(...)), "weight")`.
- **`TestObjectSpecCompile`** (e2e):
  - `test_attr_tensor_dim_dynamic` — plain Python container, `obj.w + 1`, asserts `SymInt` at dim 0 and no recompile across dim-0 changes.
  - `test_nn_module_parameter_dim_dynamic` — realistic case: `nn.Parameter` reached via `self.weight`. Exercises `NNModuleSource` unwrap, `UnspecializedParamBufferSource` collapse, module-wrapping pre-mark skip, and `wrap_tensor` spec-aware bypasses.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: torch/fx/experimental/dynamic_spec.py [hunk 1]
```diff
@@ -161,10 +163,140 @@ def to_jsonable(self) -> dict[str, Any]:
         }
 
 
+class ObjectSpec:
+    """Spec for any Python object's attributes.
+
+    Models attribute access on a Python object (e.g., an ``nn.Module``'s
+    parameters / buffers / submodules; ``self`` inside an instance
+    method). Each field corresponds to an ``AttrSource`` in the dynamo
```
Comment (laithsakka) [inline torch/fx/experimental/dynamic_spec.py]: or user defined objects.

File: torch/fx/experimental/dynamic_spec.py [hunk 2]
```diff
@@ -161,10 +163,140 @@ def to_jsonable(self) -> dict[str, Any]:
...
+    builder — when ``model.weight`` is traced, the spec walk descends
```
Comment (laithsakka) [inline torch/fx/experimental/dynamic_spec.py]: ```— when ``model.weight`` is traced, the spec walk descends
    via ``ObjectSpec._fields["weight"]``
```
? what do you mean

File: torch/fx/experimental/dynamic_spec.py [hunk 3]
```diff
@@ -161,10 +163,140 @@ def to_jsonable(self) -> dict[str, Any]:
...
+                name: spec.to_jsonable() if hasattr(spec, "to_jsonable") else spec
+                for name, spec in self._fields.items()
+            },
+        }
+
+
+class DictSpec:
+    """Spec for a Python ``dict``-typed value.
+
+    Models dict-subscript access (``d["x"]``) on a value — the value
```
Comment (laithsakka) [inline torch/fx/experimental/dynamic_spec.py]: value i any IntermediateSpec
what do you mean by 

"the value might be a top-level function arg that's a dict, or a dict-typed
attribute reached through an ``ObjectSpec`` (``obj.config["batch"]``)."

File: torch/fx/experimental/dynamic_spec.py [hunk 4]
```diff
@@ -161,10 +163,140 @@ def to_jsonable(self) -> dict[str, Any]:
...
+    might be a top-level function arg that's a dict, or a dict-typed
+    attribute reached through an ``ObjectSpec`` (``obj.config["batch"]``).
+    Each entry corresponds to a ``DictGetItemSource`` (or
```
Comment (laithsakka) [inline torch/fx/experimental/dynamic_spec.py]: no need to mention how dynamo source represents the spec here. 
since specs are not only for dynamo

File: torch/fx/experimental/dynamic_spec.py (hunk 1 above)
Comment (laithsakka) [reply torch/fx/experimental/dynamic_spec.py]: no need to mention how dynamo represent them here.

File: torch/fx/experimental/dynamic_spec.py [hunk 5]
```diff
@@ -161,10 +163,140 @@ def to_jsonable(self) -> dict[str, Any]:
...
+    ``GetItemSource`` with a ``str`` index) in the dynamo builder.
+
+    Distinct from ``ParamsSpec``: ``ParamsSpec`` describes the
```
Comment (laithsakka) [inline torch/fx/experimental/dynamic_spec.py]: i mean its indeed distinct it represent a dictionary in the user code.
you might want to mention what restrictions on the keys, (integer and strings or what exactly?) did you check what keys dynamo supports.

File: torch/fx/experimental/dynamic_spec.py [hunk 6]
```diff
@@ -161,10 +163,140 @@ def to_jsonable(self) -> dict[str, Any]:
...
+    *function's named arguments* (signature-level identifiers);
+    ``DictSpec`` describes a runtime ``dict`` value's keys.
+
+    Constructor::
+
+        DictSpec({key: IntermediateSpec, ...})
+
+    Values may be leaves (``TensorSpec`` / ``IntVar`` / ``int`` /
```
Comment (laithsakka) [inline torch/fx/experimental/dynamic_spec.py]: no need for over specifying  the type already says value is IntermediateSpec same for fields in object spec.

File: test/dynamo/test_dynamic_spec.py [hunk 7]
```diff
@@ -501,5 +504,268 @@ def fn(x, n):
...
 
 
+class TestObjectSpec(TestCase):
+    """``ObjectSpec`` data class — construction, access, repr."""
+
+    def setUp(self):
+        super().setUp()
+        _reset_uid_counter()
+
+    def test_empty(self):
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py]: test not needed over testing

File: test/dynamo/test_dynamic_spec.py [hunk 8]
```diff
@@ -501,5 +504,268 @@ def fn(x, n):
...
+        """An ``ObjectSpec`` constructed without args has zero fields."""
+        os = ObjectSpec()
+        self.assertEqual(len(os), 0)
+
+    def test_dict_construction(self):
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py:512]: do we really need this test?

File: test/dynamo/test_dynamic_spec.py [hunk 9]
```diff
@@ -501,5 +504,268 @@ def fn(x, n):
...
+        self.assertIs(
+            lookup_spec_from_dynamo_source(self._local("x"), shapes_spec_leaf),
+            leaf,
+        )
+        # Container at the top-level: bare LocalSource without a path
+        # has no leaf to apply, so returns None.
+        shapes_spec_obj = ShapesSpec(
+            params=ParamsSpec({"model": ObjectSpec({"weight": leaf})})
+        )
+        self.assertIsNone(
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py]: this shall be an error , we shall not call lookup_spec_from_dynamo_source to extract an object spec we only call it to extract tensor spec or intspec or none
or just do not test this path that is never excersized.

File: test/dynamo/test_dynamic_spec.py [hunk 10]
```diff
@@ -501,5 +504,268 @@ def fn(x, n):
...
+        _reset_uid_counter()
+
+    def _local(self, name):
+        return LocalSource(name, is_input=True)
+
+    def _attr(self, base, member):
+        return AttrSource(base, member)
+
+    def test_local_source_root_returns_leaf_spec(self):
+        """Bare ``LocalSource`` resolves to its top-level entry only when
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py]: comment is not clear code is easier to follow lol just remove it maybe?

File: test/dynamo/test_dynamic_spec.py [hunk 11]
```diff
@@ -501,5 +504,268 @@ def fn(x, n):
...
+            )
+        )
+        self.assertIsNone(
+            lookup_spec_from_dynamo_source(
+                self._attr(self._local("model"), "bias"), shapes_spec
+            )
+        )
+
+    def test_attr_against_non_objectspec_returns_none(self):
+        """Asking for an attr off a top-level non-``ObjectSpec`` returns ``None``."""
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py]: this kind of will never happen we only call lookup_spec_from_dynamo_source on tensors/ints

File: test/dynamo/test_dynamic_spec.py [hunk 12]
```diff
@@ -501,5 +504,268 @@ def fn(x, n):
...
+            backend=backend,
+            shapes_spec=ShapesSpec(
+                params=ParamsSpec(
+                    {"self": ObjectSpec({"weight": TensorSpec([ShapeVar("h"), None])})}
+                )
+            ),
+        )
+
+        compiled()
+        self.assertEqual(len(backend.graphs), 1)
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py:664]: update
 self.weight
and recompile and check
        self.assertEqual(len(backend.graphs), 1)?

File: torch/fx/experimental/dynamic_spec.py [hunk 13]
```diff
@@ -161,10 +163,120 @@ def to_jsonable(self) -> dict[str, Any]:
...
+    Example::
+
+        DictSpec({"x": TensorSpec([ShapeVar("h"), None])})
+        DictSpec({"config": DictSpec({"batch": IntVar()})})
+    """
+
+    def __init__(self, entries: dict[str, IntermediateSpec] | None = None) -> None:
+        self._entries: dict[str, IntermediateSpec] = dict(entries) if entries else {}
+
+    def __getitem__(self, key: str) -> IntermediateSpec:
```
Comment (laithsakka) [inline torch/fx/experimental/dynamic_spec.py:235]: mm this return a mutable ref?
shall we not support it? idk..

**Review by laithsakka — APPROVED** [review]

File: torch/fx/experimental/dynamic_spec.py (hunk 13 above)
Comment (fxdawnn) [reply torch/fx/experimental/dynamic_spec.py:235]: it's just get and no set. So it's immutable now... (i believe.)

Comment (fxdawnn) [reply torch/fx/experimental/dynamic_spec.py:235]: But i agree we should drop this....

File: test/dynamo/test_dynamic_spec.py (hunk 9 above)
Comment (fxdawnn) [reply test/dynamo/test_dynamic_spec.py]: fixed!

File: test/dynamo/test_dynamic_spec.py (hunk 10 above)
Comment (fxdawnn) [reply test/dynamo/test_dynamic_spec.py]: fixed!

File: test/dynamo/test_dynamic_spec.py (hunk 11 above)
Comment (fxdawnn) [reply test/dynamo/test_dynamic_spec.py]: remove the test of look_spec completely.

File: test/dynamo/test_dynamic_spec.py (hunk 7 above)
Comment (fxdawnn) [reply test/dynamo/test_dynamic_spec.py]: removed.

File: test/dynamo/test_dynamic_spec.py (hunk 12 above)
Comment (fxdawnn) [reply test/dynamo/test_dynamic_spec.py:664]: updated.
